ocaml/typing/solver.ml

```diff
@@ -0,0 +1,779 @@
+      | Amodejoin (c, mvs) -> Amodejoin (c, List.map disallow_left_mv mvs)
+      | Amodemeet (c, mvs) -> Amodemeet (c, List.map disallow_left_mv mvs)
+
+    and disallow_left_mv :
+        type a l r. (a, l * r) morphvar -> (a, disallowed * r) morphvar =
+      function
+      | Amorphvar (v, m) -> Amorphvar (v, C.disallow_left m)
+
+    let rec disallow_right :
+        type a l r. (a, l * r) mode -> (a, l * disallowed) mode = function
+      | Amode c -> Amode c
+      | Amodevar mv -> Amodevar (disallow_right_mv mv)
+      | Amodejoin (c, mvs) -> Amodejoin (c, List.map disallow_right_mv mvs)
+      | Amodemeet (c, mvs) -> Amodemeet (c, List.map disallow_right_mv mvs)
+
+    and disallow_right_mv :
+        type a l r. (a, l * r) morphvar -> (a, l * disallowed) morphvar =
+      function
+      | Amorphvar (v, m) -> Amorphvar (v, C.disallow_right m)
+  end)
+
+  let mlower dst (Amorphvar (var, morph)) = C.apply dst morph var.lower
+
+  let mupper dst (Amorphvar (var, morph)) = C.apply dst morph var.upper
+
+  let min (type a) (obj : a C.obj) = Amode (C.min obj)
+
+  let max (type a) (obj : a C.obj) = Amode (C.max obj)
+
+  let of_const a = Amode a
+
+  let apply_morphvar dst morph (Amorphvar (var, morph')) =
+    Amorphvar (var, C.compose dst morph morph')
+
+  let apply :
+      type a b l r.
+      b C.obj -> (a, b, l * r) C.morph -> (a, l * r) mode -> (b, l * r) mode =
+   fun dst morph -> function
+    | Amode a -> Amode (C.apply dst morph a)
+    | Amodevar mv -> Amodevar (apply_morphvar dst morph mv)
+    | Amodejoin (a, vs) ->
+      Amodejoin (C.apply dst morph a, List.map (apply_morphvar dst morph) vs)
+    | Amodemeet (a, vs) ->
+      Amodemeet (C.apply dst morph a, List.map (apply_morphvar dst morph) vs)
+
+  let update_lower (type a) ?log (obj : a C.obj) v a =
+    (match log with
+    | None -> ()
+    | Some log -> log := Clower (v, v.lower) :: !log);
+    v.lower <- C.join obj v.lower a;
+    if not (C.le obj v.lower v.upper)
+    then (
+      Format.eprintf "range insane after update_lower: %a\n" (print_var obj) v;
+      assert false)
+
+  let update_upper (type a) ?log (obj : a C.obj) v a =
+    (match log with
+    | None -> ()
+    | Some log -> log := Cupper (v, v.upper) :: !log);
+    v.upper <- C.meet obj v.upper a;
+    if not (C.le obj v.lower v.upper)
+    then (
+      Format.eprintf "range insane after update_lower: %a\n" (print_var obj) v;
+      assert false)
+
+  let set_vlower ?log v vlower =
+    (match log with
+    | None -> ()
+    | Some log -> log := Cvlower (v, v.vlower) :: !log);
+    v.vlower <- vlower
+
+  let submode_cv : type a. ?log:_ -> a C.obj -> a -> a var -> (unit, a) Result.t
+      =
+    fun (type a) ?log (obj : a C.obj) a' v ->
+     if C.le obj a' v.lower
+     then Ok ()
+     else if not (C.le obj a' v.upper)
+     then Error v.upper
+     else (
+       update_lower ?log obj v a';
+       if C.le obj v.upper v.lower then set_vlower ?log v [];
```

Comment:
Yeah, adding a boolean indicating this variable has been fully constrained won't make most things faster. I will add comment. 

WRT where to do the check, note that in `submode_vc` we call `update_lower` multiple times, but only check once, which is more efficient than checking once every `update_lower`.

Reply:
I'm not yet convinced here. `submode_vc` might call `update_lower` multiple times because it gets successively better information as it traverses `vlower`. That's all good. A final check between `lower` and `upper` is sufficient.

But I think we can do better: if one of the updates of `lower` means that `upper` is less than `lower`, we can just stop the iteration down `vlower` in its entirety. So this does suggest comparing `lower` against `upper` at every update -- doing so might let us short-circuit the iteration.

Reply:
We can't just short-circuit, as some of `submode_mvc` sub-call might raise errors. This is the main purpose of the loop - updating `v.lower` is just an optimisation.